src/TorchSharp/Utils/TensorAccessor.cs

```diff
@@ -158,6 +157,35 @@ public System.Array ToNDArray()
             }
         }
 
+        private Array ToNDArray(long[] shape, long[] strides)
+        {
+            Array array = Array.CreateInstance(typeof(T), shape);
+            long[] indexes = new long[_tensor.ndim];
+            long[] off = new long[_tensor.ndim];
+
+            while (true) {
+                unsafe {
+                    T* ptr = (T*)_tensor_data_ptr;
+                    array.SetValue(ptr[off[array.Rank - 1]], indexes);
+                }
+
+                for (int i = array.Rank - 1; i >= 0; i--) {
+                    if (indexes[i] < array.GetLength(i) - 1) {
+                        indexes[i]++;
+                        off[i] += strides[i];
+                        for (int j = i; j < array.Rank - 1; j++)
+                            off[j + 1] = off[j];
+                        break;
+                    } else {
+                        if (i == 0) {
+                            return array;
+                        }
+                        indexes[i] = 0;
+                    }
+                }
+            }
+        }
```

Comment:
Ah, yes of course. I missed that.

I wonder if there's a way to avoid recalculating the 'off' array for every element? The offset only changes with the rightmost dimension, and it looks like you're recalculating the whole thing every time. Maybe recursion based on the number of dimensions, and update the offset with the stride of the current dimension, then do the SetValue() at the deepest level of recursion? Just an idea -- fewer recalculations of 'off' I believe.

Reply:
Also, the `GetLength(i)` call seems to be unnecessary -- the shape of the array isn't changing, so why not just use the shape array that is passed in? Doesn't it contain the same information?

Reply:
I only recalculate the offset when there is a carry.

```
for (int j = i; j < array.Rank - 1; j++)
    off[j + 1] = off[j];
```

`j < array.Rank - 1` Therefore, when the rightmost single digit changes, it will not enter the for

Reply:
> Ah, yes of course. I missed that.
> 
> I wonder if there's a way to avoid recalculating the 'off' array for every element? The offset only changes with the rightmost dimension, and it looks like you're recalculating the whole thing every time. Maybe recursion based on the number of dimensions, and update the offset with the stride of the current dimension, then do the SetValue() at the deepest level of recursion? Just an idea -- fewer recalculations of 'off' I believe.

@NiklasGustafsson , I don't think the problem you mentioned exists.

```
indexes :
1, 0, 0 -> 1, 0, 1
It will only execute:
indexes[i]++;
off[i] += strides[i];

0, 9, 9 -> 1, 0, 0
indexes[1] and indexes [2] will execute:
indexes[i] = 0;

indexes[0] will execute:
indexes[i]++;
off[i] += strides[i];
for (int j = i; j < array.Rank - 1; j++)
    off[j + 1] = off[j];
```